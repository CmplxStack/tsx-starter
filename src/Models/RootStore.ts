import {
  types,
  Instance,
  SnapshotOut,
  SnapshotIn,
} from "mobx-state-tree";
import { Session } from "./Session";
import { SESSION_STORE } from "@Shared/Constants/Models";

const RootStore = types.model({
  [SESSION_STORE]: types.optional(Session, {}),
});
export interface IRootStoreInstance
  extends Instance<typeof RootStore> {}
export interface IRootStoreSnapshotOut
  extends SnapshotOut<typeof RootStore> {}
export interface IRootStoreSnapshotIn
  extends SnapshotIn<typeof RootStore> {}
export { RootStore };
