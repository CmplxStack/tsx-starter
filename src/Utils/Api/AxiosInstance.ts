import Axios from "axios";

const isDev = process.env.NODE_ENV === "development" || "test";

const generateAxiosInstance = () => {
  let AxiosConfig = isDev
    ? require("./Configs/MockConfig").default
    : require("./Configs/ProdConfig").default;

  let tempAxiosInstance = Axios.create(AxiosConfig);
  tempAxiosInstance.interceptors.request.use(
    (request) => {
      //something
      console.log("Request:", request);
      return request;
    },
    (error) => {
      return Promise.reject(error);
    },
  );
  tempAxiosInstance.interceptors.response.use(
    (response) => {
      console.log("Response:", response);
      return response;
    },
    (error) => {
      return Promise.reject(error);
    },
  );
  return tempAxiosInstance;
};

let CustomAxiosInstance = generateAxiosInstance();

export type ICustomAxiosInstance = typeof CustomAxiosInstance;

export { CustomAxiosInstance };
