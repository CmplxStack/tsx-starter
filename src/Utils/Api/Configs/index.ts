export * from "./ProdConfig";
export * from "./MockConfig";
