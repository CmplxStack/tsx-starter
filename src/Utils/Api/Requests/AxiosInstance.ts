import Axios, { AxiosInstance } from "axios";

const isDev = process.env.NODE_ENV === "development" || "test";

const generateAxiosInstance = () => {
  let AxiosConfig = isDev
    ? require("../Configs/MockConfig").default
    : require("../Configs/ProdConfig").default;

  let tempAxiosInstance = Axios.create(AxiosConfig);

  tempAxiosInstance.interceptors.request.use(
    (request) => {
      //something
      return request;
    },
    (error) => {
      console.log(error);
      return Promise.reject(error);
    },
  );
  tempAxiosInstance.interceptors.response.use(
    (response) => {
      //something
      if (response.status === 200) {
        return response.data;
      }
      return response;
    },
    (error) => {
      console.log(error);
      return Promise.reject(error);
    },
  );
  return tempAxiosInstance;
};

let CustomAxios: AxiosInstance = generateAxiosInstance();

export { CustomAxios };
