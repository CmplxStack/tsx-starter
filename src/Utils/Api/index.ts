import * as CustomAxiosRequests from "./Requests";

export type ICustomAxiosRequests = typeof CustomAxiosRequests;

export { CustomAxiosRequests };
