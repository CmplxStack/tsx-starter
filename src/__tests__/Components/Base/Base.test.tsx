import React from "react";
import { ReactWrapper, mount } from "enzyme";
import { createMuiTheme } from "@material-ui/core";
import { ThemeProvider } from "@material-ui/styles";
import { Provider } from "mobx-react";
import { BaseComponent } from "@Components";
import { RootStore } from "@Models/RootStore";
import { CustomAxiosRequests } from "@Utils/Api";
import { MockTodoResponse } from "@Helpers";
import {
  onPatch,
  getSnapshot,
  IJsonPatch,
  onSnapshot,
  ModelSnapshotType,
  applySnapshot,
  applyPatch,
} from "mobx-state-tree";
import Axios from "axios";
import { CustomAxiosInstance } from "@Utils/Api/AxiosInstance";

jest.mock("@Utils/Api/AxiosInstance");

let mockedAxios = CustomAxiosInstance as jest.Mocked<typeof Axios>;

describe("Base Component Element Tests", () => {
  let fullWrapper: ReactWrapper;
  let baseComponent: ReactWrapper;

  const ROOT_STORE = RootStore.create({}, { CustomAxiosRequests });
  const theme = createMuiTheme({});

  let RootComponent = () => (
    <Provider {...ROOT_STORE}>
      <ThemeProvider theme={theme}>
        <BaseComponent />
      </ThemeProvider>
    </Provider>
  );
  beforeAll(() => {
    fullWrapper = mount(<RootComponent />);
    baseComponent = fullWrapper.find("#base-component-el");
  });
  afterAll(() => {
    fullWrapper.unmount();
  });
  test("should render parent component with id `base-component-el", () => {
    expect(baseComponent).toExist();
  });

  test("should render a button element with id `gen-todo-btn`", () => {
    let genTodoButton = baseComponent.find("#gen-todo-button");
    expect(genTodoButton).toExist();
  });
  test("should render a message element with id `message-element`", () => {
    let messageElement = baseComponent.find("#message-el");
    expect(messageElement).toExist();
  });
});

describe("Base tests", () => {
  let fullWrapper: ReactWrapper;
  mockedAxios.get.mockResolvedValue({ data: MockTodoResponse });
  let axiosGetSpy = jest.spyOn(CustomAxiosInstance, "get");

  test("should have SESSION_STORE and classes as props", () => {
    const ROOT_STORE = RootStore.create({}, { CustomAxiosRequests });
    const theme = createMuiTheme({});

    let RootComponent = () => (
      <Provider {...ROOT_STORE}>
        <ThemeProvider theme={theme}>
          <BaseComponent />
        </ThemeProvider>
      </Provider>
    );
    fullWrapper = mount(<RootComponent />);
    let baseComponent = fullWrapper.find("BaseComponent");
    expect(baseComponent).toExist();
    expect(baseComponent).toHaveProp("SESSION_STORE");
    expect(baseComponent).toHaveProp("classes");
    fullWrapper.unmount();
  });

  test("Button onClick calls Axios and updates mobx state", async () => {
    let patches: IJsonPatch[] = [];
    let snapshots: ModelSnapshotType<any>[] = [];

    const ROOT_STORE = RootStore.create({}, { CustomAxiosRequests });
    const theme = createMuiTheme({});

    let RootComponent = () => (
      <Provider {...ROOT_STORE}>
        <ThemeProvider theme={theme}>
          <BaseComponent />
        </ThemeProvider>
      </Provider>
    );

    fullWrapper = mount(<RootComponent />);

    onSnapshot(ROOT_STORE, (snapshot) => {
      snapshots.push(snapshot);
    });
    onPatch(ROOT_STORE, (patch) => {
      patches.push(patch);
    });

    let baseComponent = fullWrapper.find("BaseComponent");
    expect(baseComponent).toExist();
    let genTodoButton = await baseComponent.find("#gen-todo-button").find("button");
    expect(genTodoButton).toExist();
    let messageElement = baseComponent.find("#message-el");
    expect(messageElement).toExist();
    expect(messageElement).toHaveText("Welcome!");
    await genTodoButton.simulate("click");

    await expect(axiosGetSpy).toHaveBeenCalled();

    await getSnapshot(ROOT_STORE);
    await expect(ROOT_STORE).toMatchSnapshot();
    // or.....
    // await expect(await getSnapshot(ROOT_STORE)).toMatchSnapshot();
    expect(patches).toMatchSnapshot();
    expect(snapshots).toMatchSnapshot();
    // must be after expect(ROOT_STORE).toMatchSnapshot();
    expect(messageElement).toHaveText(MockTodoResponse.title);

    fullWrapper.unmount();
  });
});
