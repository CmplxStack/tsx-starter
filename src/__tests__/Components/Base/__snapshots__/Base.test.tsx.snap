// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base tests Button onClick calls Axios and updates mobx state 1`] = `
Object {
  "SESSION_STORE": Object {
    "isLoading": "pending",
    "sessionName": undefined,
    "welcomeMessage": "Welcome!",
  },
}
`;

exports[`Base tests Button onClick calls Axios and updates mobx state 2`] = `
Array [
  Object {
    "op": "replace",
    "path": "/SESSION_STORE/welcomeMessage",
    "value": "Mock Todo Title",
  },
  Object {
    "op": "replace",
    "path": "/SESSION_STORE/isLoading",
    "value": "done",
  },
]
`;

exports[`Base tests Button onClick calls Axios and updates mobx state 3`] = `
Array [
  Object {
    "SESSION_STORE": Object {
      "isLoading": "done",
      "sessionName": undefined,
      "welcomeMessage": "Mock Todo Title",
    },
  },
]
`;
